packages/myst-execute/src/execute.ts

```diff
@@ -0,0 +1,345 @@
+  nodes: (ICellBlock | InlineExpression)[],
+  vfile: VFile,
+): Promise<{
+  results: (IOutput[] | IExpressionResult)[];
+  errorOccurred: boolean;
+}> {
+  let errorOccurred = false;
+
+  const results: (IOutput[] | IExpressionResult)[] = [];
+  for (const matchedNode of nodes) {
+    if (isCellBlock(matchedNode)) {
+      // Pull out code to execute
+      const code = select('code', matchedNode) as Code;
+      const { status, outputs } = await executeCode(kernel, code.value);
+      // Cache result
+      results.push(outputs);
+
+      // Check for errors
+      const metadata = matchedNode.data || {};
+      const allowErrors = !!metadata?.tags?.['raises-exception'];
+      if (status === 'error' && !allowErrors) {
+        fileWarn(vfile, 'An exception occurred during code execution, halting further execution');
+        // Make a note of the failure
+        errorOccurred = true;
+        break;
+      }
+    } else if (isInlineExpression(matchedNode)) {
+      // Directly evaluate the expression
+      const { status, result } = await evaluateExpression(kernel, matchedNode.value);
+
+      // Check for errors
+      if (status === 'error') {
+        fileWarn(
+          vfile,
+          'An exception occurred during expression evaluation, halting further execution',
+        );
+        // Make a note of the failure
+        errorOccurred = true;
+        break;
+      }
+      // Cache result
+      results.push(result);
+    } else {
+      assert(false);
+    }
+  }
+  return { results, errorOccurred };
+}
+
+/**
+ * Apply computed outputs (MIME bundles, stdout, etc.) to MDAST.
+ * If an unanticipated error occurred during execution, the length of
+ * `computedResult` may not correspond to that of `nodes` (it may be shorter)
+ *
+ * @param nodes executable MDAST nodes
+ * @param computedResult computed results for each node
+ */
+function applyComputedOutputsToNodes(
+  nodes: (ICellBlock | InlineExpression)[],
+  computedResult: (IOutput[] | IExpressionResult)[],
+) {
+  for (const matchedNode of nodes) {
+    // Pull out the result for this node
+    const thisResult = computedResult.shift();
+
+    if (isCellBlock(matchedNode)) {
+      // Pull out output to set data
+      const output = select('output', matchedNode) as unknown as { data: IOutput[] };
+      // Set the output array to empty if we don't have a result (e.g. due to a kernel error)
+      output.data = thisResult === undefined ? [] : (thisResult as IOutput[]);
+    } else if (isInlineExpression(matchedNode)) {
+      // Set data of expression to the result, or empty if we don't have one
+      matchedNode.result = // TODO: FIXME .data
+        thisResult === undefined ? undefined : (thisResult as unknown as Record<string, unknown>);
+    } else {
+      // This should never happen
+      assert(false);
+    }
+  }
+}
+
+export type Options = {
+  cache: ICache<(IExpressionResult | IOutput[])[]>;
+  sessionFactory: () => Promise<SessionManager | undefined>;
```

Comment:
This maybe should return `SessionManager`, but I don't have a feel for MyST's error handling yet, so erring on the side of "report this in the transform".

Reply:
Any error handling should be done through `fileError` and `fileWarning` and passing around a vfile.

Something like:
https://github.com/executablebooks/mystjs/blob/main/packages/myst-cli/src/transforms/code.ts#L159

This will likely also be adding a new `RuleId` to keep track of these for later processing/ignoring/filtering.